core/drivers/crypto/se050/core/ecc.c

```diff
@@ -0,0 +1,630 @@
+		return kAlgorithm_SSS_ECDSA_SHA256;
+	case TEE_ALG_ECDSA_P384:
+		return kAlgorithm_SSS_ECDSA_SHA384;
+	case TEE_ALG_ECDSA_P521:
+		return kAlgorithm_SSS_ECDSA_SHA512;
+	default:
+		EMSG("ecc curve 0x%x not enabled", algo);
+		return kAlgorithm_None;
+	}
+}
+
+static uint32_t cipher_tee2se050(uint32_t curve)
+{
+	switch (curve) {
+	case TEE_ECC_CURVE_NIST_P192:
+	case TEE_ECC_CURVE_NIST_P224:
+	case TEE_ECC_CURVE_NIST_P256:
+	case TEE_ECC_CURVE_NIST_P384:
+	case TEE_ECC_CURVE_NIST_P521:
+		return kSSS_CipherType_EC_NIST_P;
+	default:
+		EMSG("cipher 0x%x not enabled", curve);
+		return kSSS_CipherType_NONE;
+	}
+}
+
+static uint32_t curve_tee2se050(uint32_t curve)
+{
+	switch (curve) {
+	case TEE_ECC_CURVE_NIST_P192:
+		return kSE05x_ECCurve_NIST_P192;
+	case TEE_ECC_CURVE_NIST_P224:
+		return kSE05x_ECCurve_NIST_P224;
+	case TEE_ECC_CURVE_NIST_P256:
+		return kSE05x_ECCurve_NIST_P256;
+	case TEE_ECC_CURVE_NIST_P384:
+		return kSE05x_ECCurve_NIST_P384;
+	case TEE_ECC_CURVE_NIST_P521:
+		return kSE05x_ECCurve_NIST_P521;
+	default:
+		EMSG("curve 0x%x not enabled", curve);
+		return kSE05x_ECCurve_NA;
+	}
+}
+
+static uint32_t curve_se0502tee(uint32_t curve)
+{
+	switch (curve) {
+	case kSE05x_ECCurve_NIST_P192:
+		return TEE_ECC_CURVE_NIST_P192;
+	case kSE05x_ECCurve_NIST_P224:
+		return TEE_ECC_CURVE_NIST_P224;
+	case kSE05x_ECCurve_NIST_P256:
+		return TEE_ECC_CURVE_NIST_P256;
+	case kSE05x_ECCurve_NIST_P384:
+		return TEE_ECC_CURVE_NIST_P384;
+	case kSE05x_ECCurve_NIST_P521:
+		return TEE_ECC_CURVE_NIST_P521;
+	default:
+		EMSG("curve 0x%x not enabled", curve);
+		return TEE_CRYPTO_ELEMENT_NONE;
+	}
+}
+
+static bool bn_alloc_max(struct bignum **s)
+{
+	*s = crypto_bignum_allocate(4096);
+
+	return *s;
+}
+
+static TEE_Result ecc_get_key_size(uint32_t curve, uint32_t algo,
+				   size_t *bytes, size_t *bits)
+{
+	/*
+	 * Note GPv1.1 indicates TEE_ALG_ECDH_NIST_P192_DERIVE_SHARED_SECRET
+	 * but defines TEE_ALG_ECDH_P192
+	 */
+	switch (curve) {
+	case TEE_ECC_CURVE_NIST_P192:
+		*bits = 192;
+		*bytes = 24;
+		if (algo && algo != TEE_ALG_ECDSA_P192 &&
+		    algo != TEE_ALG_ECDH_P192)
+			return TEE_ERROR_BAD_PARAMETERS;
+		break;
+	case TEE_ECC_CURVE_NIST_P224:
+		*bits = 224;
+		*bytes = 28;
+		if (algo && algo != TEE_ALG_ECDSA_P224 &&
+		    algo != TEE_ALG_ECDH_P224)
+			return TEE_ERROR_BAD_PARAMETERS;
+		break;
+	case TEE_ECC_CURVE_NIST_P256:
+		*bits = 256;
+		*bytes = 32;
+		if (algo && algo != TEE_ALG_ECDSA_P256 &&
+		    algo != TEE_ALG_ECDH_P256)
+			return TEE_ERROR_BAD_PARAMETERS;
+		break;
+	case TEE_ECC_CURVE_NIST_P384:
+		*bits = 384;
+		*bytes = 48;
+		if (algo && algo != TEE_ALG_ECDSA_P384 &&
+		    algo != TEE_ALG_ECDH_P384)
+			return TEE_ERROR_BAD_PARAMETERS;
+		break;
+	case TEE_ECC_CURVE_NIST_P521:
+		*bits = 521;
+		*bytes = 66;
+		if (algo && algo != TEE_ALG_ECDSA_P521 &&
+		    algo != TEE_ALG_ECDH_P521)
+			return TEE_ERROR_BAD_PARAMETERS;
+		break;
+	default:
+		*bits = 0;
+		*bytes = 0;
```

Comment:
I don't think setting bthese to 0 is needed. It's not done for `TEE_ERROR_BAD_PARAMETERS` cases.

Reply:
ok